tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -4027,44 +4027,102 @@ Status ConvertGather(OpConverterParams* params) {
-  while (input_dim.nbDims != 3) {
-    input_dim.d[input_dim.nbDims++] = 1;
-  }
-  TF_RETURN_IF_ERROR(params->converter->PrepareTensorForShape(
-      tensor_input, input_dim, /*validation_only=*/false, &tensor));
-
+  const int noutput = weights.shape_.d[0];
   nvinfer1::IFullyConnectedLayer* layer =
       params->converter->network()->addFullyConnected(
-          *tensor, noutput, weights.GetTrtWeights(), biases.GetTrtWeights());
-  TFTRT_RETURN_ERROR_IF_NULLPTR(layer, node_name);
+          *tensor_a, noutput, weights.GetTrtWeights(), biases.GetTrtWeights());
 
+  TFTRT_RETURN_ERROR_IF_NULLPTR(layer, node_name);
   nvinfer1::ITensor* output_tensor = layer->getOutput(0);
+  params->outputs->push_back(TRT_TensorOrWeights(output_tensor));
+
+  // Reshape output to 1D - this will be a no-op unless using int8 precision.
   auto output_dim = output_tensor->getDimensions();
   output_dim.nbDims = 1;
   TF_RETURN_IF_ERROR(params->converter->PrepareTensorForShape(
       TRT_TensorOrWeights(output_tensor), output_dim, /*validation_only=*/false,
       &output_tensor));
+
+  return Status::OK();
+}
+
+Status ConvertMatMulHelper(OpConverterParams* params,
+                           TRT_TensorOrWeights input_a,
+                           TRT_TensorOrWeights input_b, bool transpose_a,
+                           bool transpose_b, string node_name) {
+  // If an FC layer can be used and would be faster, use that instead.
+  const bool should_use_fc =
+      !transpose_a && input_a.is_tensor() && input_b.is_weights() &&
+      input_a.GetTrtDims().nbDims >= 3 && input_b.GetTrtDims().nbDims == 2;
+  // If int8 is specified, FC must be used, as MM does not support int8 at this
+  // time.
+  if (should_use_fc ||
+      params->converter->precision_mode() == TrtPrecisionMode::INT8) {
+    return ConvertFullyConnectedHelper(params, input_a.tensor(), input_b.weights(),
+                           transpose_b, node_name);
+  }
+
+  constexpr auto get_matrix_op =
+      [](nvinfer1::ITensor* in, bool transpose) -> nvinfer1::MatrixOperation {
+    return (in->getDimensions().nbDims < 2)
+               ? nvinfer1::MatrixOperation::kVECTOR
+               : (transpose) ? nvinfer1::MatrixOperation::kTRANSPOSE
+                             : nvinfer1::MatrixOperation::kNONE;
+  };
+
+  // If the MatMul operand is a constant, applies transposes at conversion-time
+  // as necessary. If the operand is a tensor, does nothing. If required
+  // transposes were applied, sets transpose to false.
+  const auto prepare_matmul_operand =
+      [&params](TRT_TensorOrWeights operand,
+                bool* transpose) -> nvinfer1::ITensor* {
+    if (operand.is_tensor()) {
+      return operand.tensor();
+    } else {
+      TRT_ShapedWeights weights(operand.weights().TrtDType());
+      if (*transpose) {
+        weights = params->weight_store->GetTempWeights(operand.weights());
+        ReorderCKtoKC(operand.weights(), &weights);
```

Comment:
What if nbDims!=2?

Reply:
I've added an explicit check for that, which fail return an error if that's the case (I think a general transpose implementation would be outside the scope of this change).